apps/system/js/pin_page_system_dialog.js

```diff
@@ -0,0 +1,118 @@
+    this.controller = controller || {};
+    this.options = {};
+    this.render();
+    this.publish('created');
+  };
+
+  PinPageSystemDialog.prototype = Object.create(SystemDialog.prototype, {
+    visible: {
+      configurable: false,
+      get: function() {
+        return this._visible;
+      }
+    }
+  });
+
+  PinPageSystemDialog.prototype.customID = 'pin-page-dialog';
+
+  PinPageSystemDialog.prototype.DEBUG = false;
+
+  PinPageSystemDialog.prototype.name = 'PinPageSystemDialog';
+
+  PinPageSystemDialog.prototype.view = function spd_view() {
+    return `<div id="${this.instanceID}" role="dialog"
+           class="generic-dialog" data-z-index-level="system-dialog" hidden>
+              <section role="region">
+                <gaia-header id="pin-page-header" action="close">
+                  <h1 id="pin-page-title" data-l10n-id="pinning-pin-page"></h1>
+                </gaia-header>
+                <div class="container">
+                  <div id="pin-card-container"></div>
+                  <p id="pin-page-url" class="url" dir="ltr"></p>
+                  <button data-l10n-id="pinning-pin" data-action="pin"
+                  class="pin-button">
+                    Pin
+                  </button>
+                </div>
+              </section>
+            </div>`;
+  };
+
+  PinPageSystemDialog.prototype.onHide = function() {
+    this._visible = false;
+    this.controller.onHide && this.controller.onHide();
+  };
+
+  PinPageSystemDialog.prototype._registerEvents = function() {
+    this.header.addEventListener('action', this.close.bind(this));
+    this.pinButton.addEventListener('click', this.save.bind(this));
+  };
+
+  PinPageSystemDialog.prototype._fetchElements = function spl_initElements() {
+    var prefix = '#' + this.instanceID;
+    this.element = document.querySelector(prefix);
+    this.pinURL = document.getElementById('pin-page-url');
+    this.pinCardContainer = document.getElementById('pin-card-container');
+    this.header = document.querySelector(prefix + ' gaia-header');
+    var pinSelector = prefix + ' button[data-action="pin"]';
+    this.pinButton = document.querySelector(pinSelector);
+  };
+
+  PinPageSystemDialog.prototype._renderPinCard = function renderPinCard(data) {
+    this.data = data;
+    this.card = new GaiaPinCard();
+    this.card.title = data.title;
+    if (data.icon) {
+      this.card.icon = 'url(' + URL.createObjectURL(data.icon) + ')';
+    }
+
+    this.pinCardContainer.innerHTML = '';
+
+    var screenshot = data.screenshot || null;
+    var screenshotURL = screenshot ? URL.createObjectURL(screenshot) : null;
+    this.card.background = {
+      src: screenshotURL,
+      themeColor: data.theme
+    };
+    this.pinCardContainer.appendChild(this.card);
+  };
+
+  PinPageSystemDialog.prototype.save = function() {
+    var activeApp = Service.query('getTopMostWindow');
+    activeApp.appChrome.pinPage && activeApp.appChrome.pinPage();
```

Comment:
The latest spec says to show a confirmation the page has been pinned, can we squeeze this into this patch? Maybe in app_chrome.js? This is more important now the browser chrome does not collapse for pinned pages.